go-controller/pkg/node/gateway.go

```diff
@@ -408,51 +416,110 @@ func (b *bridgeConfiguration) updateInterfaceIPAddresses() ([]*net.IPNet, error)
+		return nil, err
+	}
+
+	macAddress, err := util.GetOVSPortMACAddress(uplink)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get MAC address for ovs port %s", uplink)
+	}
+
+	return &bridgeConfiguration{
+		bridgeName: bridgeName,
+		bypassRep:  bypassRep,
+		uplinkName: uplink,
+		macAddress: macAddress,
+	}, nil
+}
+
+func bridgeForRegularInterface(iface string) (*bridgeConfiguration, string, error) {
+	var bridgeName, uplinkName string
+	var err error
+	gwIntf := iface
+
+	if bridgeName, _, err = util.RunOVSVsctl("port-to-br", iface); err == nil {
+		// This is port of an OVS bridge
+		if uplinkName, err = util.GetNicName(bridgeName); err != nil {
+			return nil, "", errors.Wrapf(err, "Failed to find nic name for bridge %s", bridgeName)
+		}
+	} else if _, _, err = util.RunOVSVsctl("br-exists", iface); err != nil {
+		// This is not an OVS bridge or bridge doesn't exist. We need to create it
 		// and add cluster.GatewayIntf as a port of that bridge.
-		bridgeName, err := util.NicToBridge(intfName)
-		if err != nil {
-			return nil, errors.Wrapf(err, "NicToBridge failed for %s", intfName)
+		if bridgeName, err = util.NicToBridge(iface); err != nil {
+			return nil, "", errors.Wrapf(err, "NicToBridge failed for %s", iface)
 		}
-		res.bridgeName = bridgeName
-		res.uplinkName = intfName
+		uplinkName = iface
 		gwIntf = bridgeName
-	} else {
+	} else if uplinkName, err = getIntfName(iface); err == nil {
 		// gateway interface is an OVS bridge
-		uplinkName, err := getIntfName(intfName)
-		if err != nil {
-			return nil, errors.Wrapf(err, "Failed to find intfName for %s", intfName)
-		}
-		res.bridgeName = intfName
-		res.uplinkName = uplinkName
+		bridgeName = iface
+	} else {
+		return nil, "", errors.Wrapf(err, "Failed to find uplink name for %s", iface)
+	}
+
+	macAddress, err := util.GetOVSPortMACAddress(gwIntf)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "failed to get MAC address for ovs port %s", gwIntf)
 	}
+
+	return &bridgeConfiguration{
+		bridgeName: bridgeName,
+		uplinkName: uplinkName,
+		macAddress: macAddress,
+	}, gwIntf, nil
+}
+
+func getRepresentorAndUplinkName(intfName string) (string, string, error) {
+	deviceID, err := util.GetDeviceIDFromNetdevice(intfName)
+	if err != nil {
+		return "", "", err
+	}
+
+	return util.GetFunctionRepresentorName(deviceID)
+}
+
+func bridgeForInterface(intfName, nodeName, physicalNetworkName string, kubeNodeIP net.IP) (*bridgeConfiguration, error) {
+	var res *bridgeConfiguration
+	var gwIntf, rep, uplink string
 	var err error
-	// Now, we get IP addresses for the bridge
+
+	// Try to get representor and uplink name for the specified device.
+	// If function had succeed, then it is either switchdev VF or SF, hence, configure it as a bypass interface.
+	// If failed - falback to regular interface configuration.
+	rep, uplink, err = getRepresentorAndUplinkName(intfName)
```

Comment:
I am not sure if using this function failure is the best way to determine whether we have a Bypass Port or regular interface. I understand you are keeping the existing argument for the API, but have you considered adding a new argument for this feature to make the separation clearer (possibly)?

Reply:
> I am not sure if using this function failure is the best way

This function may succeed only for switchdev VF or SF. An alternative way could be getting port flavour. But. Respective function requires representor name:
`func GetRepresentorPortFlavour(netdev string) (PortFlavour, error) {`
..so we need to get representor anyway.

> have you considered adding a new argument for this feature

What do You mean? Adding something like `--gateway-bypass-interface`? Internally we agreed that introducing another argument looks like a duplication, because in fact we telling ovnkube to use *THIS* netdevice as a gateway interface, and we already have an argument for a gateway interface.

Reply:
> > I am not sure if using this function failure is the best way
> 
> This function may succeed only for switchdev VF or SF. An alternative way could be getting port flavour. But. Respective function requires representor name: `func GetRepresentorPortFlavour(netdev string) (PortFlavour, error) {` ..so we need to get representor anyway.
> 
I am just afraid, the error is an actual error when the intention is to use a bypass interface. I.e. Actual Error vs. Expected Error is always confusing and hard to debug.  Just my opinion here.
> > have you considered adding a new argument for this feature
> 
> What do You mean? Adding something like `--gateway-bypass-interface`? Internally we agreed that introducing another argument looks like a duplication, because in fact we telling ovnkube to use _THIS_ netdevice as a gateway interface, and we already have an argument for a gateway interface.

The confusing thing (when trying to read the code) without your slides or diagrams. It seems that there is a new feature called Bypass Interface/Port, but there is no clear way of configuring this. But rather it is indirectly from an existing input parameter. This is just my third-party perspective on it. But I agree with you, we shouldn't duplicate parameters when they do the same thing. I am not sure what the solution would be, but maybe updating the config.go:OVNGatewayFlags definition of gateway-interface may help. Adding Bypass Port inside the help text could help the user understand this feature.

Reply:
> > > I am not sure if using this function failure is the best way
> > 
> > 
> > This function may succeed only for switchdev VF or SF. An alternative way could be getting port flavour. But. Respective function requires representor name: `func GetRepresentorPortFlavour(netdev string) (PortFlavour, error) {` ..so we need to get representor anyway.
> 
> I am just afraid, the error is an actual error when the intention is to use a bypass interface. I.e. Actual Error vs. Expected Error is always confusing and hard to debug. Just my opinion here.

That is why it is logged. If You think that need to invest more time to handle this in another way, I'll do.

> 
> > > have you considered adding a new argument for this feature
> > 
> > 
> > What do You mean? Adding something like `--gateway-bypass-interface`? Internally we agreed that introducing another argument looks like a duplication, because in fact we telling ovnkube to use _THIS_ netdevice as a gateway interface, and we already have an argument for a gateway interface.
> 
> The confusing thing (when trying to read the code) without your slides or diagrams. It seems that there is a new feature called Bypass Interface/Port, but there is no clear way of configuring this. But rather it is indirectly from an existing input parameter. This is just my third-party perspective on it. But I agree with you, we shouldn't duplicate parameters when they do the same thing. I am not sure what the solution would be, but maybe updating the config.go:OVNGatewayFlags definition of gateway-interface may help. Adding Bypass Port inside the help text could help the user understand this feature.

Yes, this should be documented.

Reply:
I see that it is logged below, but we don't "fail". I guess this is ok.